pkg/shell/indexes.js

```diff
@@ -124,6 +125,66 @@ function MachinesIndex(index_options, machines, loader, mdialogs) {
             index.preload_frames(m, m.manifests);
     }
 
+    function selectActiveItem() {
+        const cur = document.activeElement;
+        if (cur.nodeName === "INPUT") {
+            const el = document.querySelector("#host-apps li:first-of-type > a");
+            if (el)
+                el.click();
+        } else {
+            cur.click();
+        }
+    }
+
+    function moveItem(nth_of_type, sibling) {
+        const cur = document.activeElement;
+        if (cur.nodeName === "INPUT") {
+            const item = document.querySelector("#host-apps li:" + nth_of_type + " > a");
+            if (item)
+                item.focus();
+        } else {
+            const next = cur.parentNode[sibling];
+            if (next)
+                next.children[0].focus();
+            else
+                document.getElementById("filter-menus").focus();
+        }
+    }
+
+    function navigate_apps(ev) {
+        if (ev.keyCode === 13) // Enter
+            selectActiveItem();
+        else if (ev.keyCode === 40) // Arrow Down
+            moveItem("first-of-type", "nextSibling");
+        else if (ev.keyCode === 38) // Arrow Up
+            moveItem("last-of-type", "previousSibling");
+        else if (ev.keyCode === 27) { // Escape - clean selection
+            document.getElementById("filter-menus").value = "";
+            update_sidebar();
+            document.getElementById("filter-menus").focus();
+        } else {
+            return false;
+        }
+        return true;
+    }
+
+    function manage_search(ev) {
+        if (!navigate_apps(ev)) {
+            if (filter_timer)
+                window.clearTimeout(filter_timer);
+
+            filter_timer = window.setTimeout(function () {
+                if (document.getElementById("filter-menus") === document.activeElement)
+                    update_sidebar();
+                filter_timer = null;
+            }, 250);
```

Comment:
It's fine to implement this by hand, but just FYI: several other pages use [throttle-debounce](https://www.npmjs.com/package/throttle-debounce), which makes this a bit easier.

Reply:
:) I just did the same that we are already doing on the same page (line 72).